### PR TITLE
Username field and FCM

### PR DESCRIPTION
**Pull request by vnagel**
Now checks if username is already taken when registering. If it is, it takes the user back to the setup profile screen to pick a different username.
Client now updates firebase instance id's on backend so it can send notifications to specific users.
Everything other than the sendRegistrationToServer() implementation of the MyFirebaseMessagingService class was copied from the [android quickstart guide](https://github.com/firebase/quickstart-android/blob/master/messaging/app/src/main/java/com/google/firebase/quickstart/fcm/java/MyFirebaseMessagingService.java). The copied content isn't necessarily correct for our app, but should be useful as an example.
Also did some refactoring.

### CODE REVIEW

File: app/src/main/java/org/michiganhackers/photoassassin/LoginPages/RegistrationActivity.java [hunk 1]
```diff
@@ -167,29 +172,84 @@ private void initializeUser(final String userId) {
...
+                                                Log.e(TAG, "Failed to add user profile pic to storage", e);
+
+                                            }
+                                        });
+                                Intent intent = new Intent(RegistrationActivity.this, MainActivity.class);
+                                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+                                startActivity(intent);
+                            } else if (errorCode.equals(Constants.ErrorCode.DUPLICATE_USERNAME)) {
+                                Log.w(TAG, "addUser returned DUPLICATE_USERNAME");
+                               handleDuplicateUsername();
```
Comment (rckhzouz) [inline app/src/main/java/org/michiganhackers/photoassassin/LoginPages/RegistrationActivity.java:210]: One suggestion is to handle duplicate usernames when the continue button is clicked in SetupProfileActivity. That way the user will know the username is taken without needing to input an email and password in the following registration activity screen.

**Review by rckhzouz — COMMENTED** [review]
This looks mostly good but I found a bug. Registering a new user works, and attempting to register a duplicate username works as intended. The bug occurs when registering a valid username after attempting to register a duplicate username and returning to the registration screen. The valid user does, however, become registered. The problem is that the user is shown an error, suggesting an invalid username, and is returned to set up a new username, but they have already successfully registered and logged in.

I also added a comment about a minor suggestion, although it is fully functional without it.

File: app/src/main/java/org/michiganhackers/photoassassin/LoginPages/RegistrationActivity.java (hunk 1 above)
Comment (vnagel) [reply app/src/main/java/org/michiganhackers/photoassassin/LoginPages/RegistrationActivity.java:210]: Okay, I can check for duplicate usernames in the setup profile activity in addition to the registration activity. The problem with just checking in the setup profile activity is that the user isn't actually created until the registration activity, and so the username could have been taken by someone else at that point. Thomas and I talked about a solution which almost always just requires checking in the setup profile activity by granting a lease for the username for a short amount of time [here](https://michiganhackers.slack.com/archives/CN9GYUF60/p1574298112000700), though we agreed that for v1 this lease solution isn't necessary.

**Comment by vnagel** [comment]
> This looks mostly good but I found a bug. Registering a new user works, and attempting to register a duplicate username works as intended. The bug occurs when registering a valid username after attempting to register a duplicate username and returning to the registration screen. The valid user does, however, become registered. The problem is that the user is shown an error, suggesting an invalid username, and is returned to set up a new username, but they have already successfully registered and logged in.

I'm unable to duplicate this bug. I did the following and it seemed to all work correctly:
1) registered a user with username "unamet1" and email "unamet1@gmail.com"
2) attempted to register a user with username "unamet1" and email "unamet2@gmail.com". I was taken back to the setup profile screen and shown an error.
3) changed the username to "unamet2" and completed the registration process normally without any errors.

**Comment by DavidJRWang** [comment]
I used the same steps as Vincent (with different names) and was also unable to duplicate the bug. The second user was able to go through to the main menu after trying an available username. We've decided to go ahead and merge.

**Pull request merged**